conda_forge_tick/git_utils.py

```diff
@@ -88,77 +97,533 @@ def github_client() -> github.Github:
+        base_branch: str = "main",
+    ):
+        """
+        Convenience method to do the following:
+        1. Clone the repository at origin_url into target_dir (resetting the directory if it already exists).
+        2. Add a remote named "upstream" with the URL upstream_url (ignoring if it already exists).
+        3. Fetch all changes from all remotes.
+        4. Checkout the base branch.
+        5. Create a new branch from the base branch with the name new_branch.
+
+        This is usually used to create a new branch for a pull request. In this case, origin_url is the URL of the
+        user's fork, and upstream_url is the URL of the upstream repository.
+
+        :param origin_url: The URL of the repository (fork) to clone.
+        :param target_dir: The directory to clone the repository into.
+        :param upstream_url: The URL of the upstream repository.
+        :param new_branch: The name of the branch to create.
+        :param base_branch: The name of the base branch to branch from.
+
+        :raises GitCliError: If a git command fails.
+        """
+        self.clone_repo(origin_url, target_dir)
+
+        try:
+            self.add_remote(target_dir, "upstream", upstream_url)
+        except GitCliError as e:
+            logger.debug(
+                "It looks like remote 'upstream' already exists. Ignoring.", exc_info=e
+            )
+            pass
+
+        self.fetch_all(target_dir)
+
+        if self.does_branch_exist(target_dir, base_branch):
+            self.checkout_branch(target_dir, base_branch)
+        else:
+            try:
+                self.checkout_branch(target_dir, f"upstream/{base_branch}", track=True)
+            except GitCliError as e:
+                logger.debug(
+                    "Could not check out with git checkout --track. Trying git checkout -b.",
+                    exc_info=e,
+                )
+
+                # not sure why this is needed, but it was in the original code
+                self.checkout_new_branch(
+                    target_dir,
+                    base_branch,
+                    start_point=f"upstream/{base_branch}",
+                )
+
+        # not sure why this is needed, but it was in the original code
+        self.reset_hard(target_dir, f"upstream/{base_branch}")
+
+        try:
+            logger.debug(
+                f"Trying to checkout branch {new_branch} without creating a new branch"
+            )
+            self.checkout_branch(target_dir, new_branch)
+        except GitCliError:
+            logger.debug(
+                f"It seems branch {new_branch} does not exist. Creating it.",
+            )
+            self.checkout_new_branch(target_dir, new_branch, start_point=base_branch)
+
+
+class GitPlatformBackend(ABC):
     """
-    gh = github3_client()
+    A backend for interacting with a git platform (e.g. GitHub).
+    Note that this class is not thread-safe, you should create a new instance for each thread.
```

Comment:
I think the lack of thread-safety is related to git and the backend more than the class in and of itself. In other words, making git commands from multiple threads on the same repo is likely not a good idea either. Similarly, making multiple API calls to GitHub trying to fork a repo from multiple threads is not a great idea, though it may in fact work. 

Should we protect calls to methods on this class (and its non-abstract children) and the class above via an RLock (https://docs.python.org/3/library/threading.html#rlock-objects) or some similar scheme?

Reply:
Good point (and my comment in the code is indeed wrong). An ever better solution is probably per-repository locking for the CLI as well as for GitHub. I will try to come up with a good solution for it.

Reply:
Ok. We don't need this to be too complex. The bot doesn't normally checkout more than one repo at a time and a single global lock would likely be ok.

Reply:
The executors.py file has a set of global locks used to ensure the hdel operations in the lazy json backends don't cause an error. Maybe use those?

Reply:
Actually looking at the code, you have to use those since they cover multiprocessing, dask, and python threads.

Reply:
You're right that a single global lock is probably sufficient for now - however I noticed that TRLOCK (thread-scoped) and PRLOCK (process-scoped) are reentrant locks while DLOCK (for dask) is not, which would make implementation a little bit more complex.

Since we don't use dask anyway, are you fine with removing it in a separate PR?

Reply:
We've used dask in the past and should keep it. Maybe dask provides a reentrant lock?

Reply:
I bet we could make the dask lock reentrant easily enough by subclassing and overriding the acquire and release methods to test on self._locked in the right way before calling the parent methods. I'll make a pr later.

Reply:
OK dask locking is fixed up and tested!

Reply:
I added it to our git methods